python/python/bystro/ancestry/ancestry_types.py

```diff
@@ -0,0 +1,158 @@
+    CLM: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    ESN: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    FIN: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    GBR: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    GIH: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    GWD: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    IBS: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    ITU: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    JPT: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    KHV: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    LWK: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    MAG: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    MSL: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    MXL: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    PEL: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    PJL: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    PUR: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    STU: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    TSI: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    YRI: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+
+
+@attr.s(kw_only=True, frozen=True)
+class SuperpopVector:
+    """A vector of probability intervals for superpopulations."""
+
+    AFR: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    AMR: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    EAS: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    EUR: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+    SAS: ProbabilityInterval = attr.field(validator=ProbIntValidator)
+
+
+@attr.s(kw_only=True, frozen=True)
+class AncestryResult:
+    """An ancestry result from a sample."""
+
+    sample_id: str = field(validator=instance_of(str))
+    populations: PopulationVector = field(validator=instance_of(PopulationVector))
+    superpops: SuperpopVector = field(validator=instance_of(SuperpopVector))
+    missingness: float = field(validator=unit_float_validator)
+
+
+@attr.s(kw_only=True, frozen=True)
+class AncestryResponse:
+    """An outgoing response from the ancestry worker."""
+
+    vcf_path: str = attr.field(validator=_vcf_validator)
+    results: list[AncestryResult] = field(validator=instance_of(list))
+
+    @results.validator
+    def _is_list_of_ancestry_results(
+        self, _attribute: attr.Attribute, values: list[AncestryResult]
+    ) -> None:
+        for i, value in enumerate(values):
+            if not isinstance(value, AncestryResult):
+                err_msg = f"Expecting list of AncestryResults, at position {i} got {value} instead"
+                raise TypeError(err_msg)
+
+    def __attrs_post_init__(self) -> None:
```

Comment:
since this also pertains to result validation (at least I think so, could we just fold into is_list_of_ancestry_results?

```
  @results.validator
    def _is_list_of_ancestry_results(
        self, _attribute: attr.Attribute, values: list[AncestryResult]
    ) -> None:
        unique_ids = set()
        for i, value in enumerate(values):
            if not isinstance(value, AncestryResult):
                err_msg = f"Expecting list of AncestryResults, at position {i} got {value} instead"
                raise TypeError(err_msg)
           if value.sample_id in unique_ids:
                raise ValueError(err_msg)
           unique_ids.add(value.sample_id)
```

Reply:
yeah I think the grand strategy for doing type- vs value validation could be a style guide topic, but can do here.